cpp/src/arrow/compute/kernels/vector_cumulative_ops.cc

```diff
@@ -0,0 +1,212 @@
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const auto& options = CumulativeOptionsWrapper<OptionsType>::Get(ctx);
+    auto start = UnboxScalar<OutType>::Unbox(*(options.start));
+    auto skip_nulls = options.skip_nulls;
+    bool encountered_null = false;
+
+    std::shared_ptr<ArrayData> out_arr;
+    NumericBuilder<OutType> builder;
+
+    switch (batch[0].kind()) {
+      case Datum::SCALAR: {
+        auto in_value = UnboxScalar<OutType>::Unbox(*(batch[0].scalar()));
+        RETURN_NOT_OK(builder.Append(start + in_value));
+        break;
+      }
+      case Datum::ARRAY: {
+        auto input = batch[0].array();
+
+        RETURN_NOT_OK(Call(ctx, *input, builder, &start, skip_nulls, &encountered_null));
+        break;
+      }
+      case Datum::CHUNKED_ARRAY: {
+        const auto& input = batch[0].chunked_array();
+
+        for (const auto& chunk : input->chunks()) {
+          RETURN_NOT_OK(
+              Call(ctx, *chunk->data(), builder, &start, skip_nulls, &encountered_null));
+        }
+        break;
+      }
+      default:
+        return Status::NotImplemented(
+            "Unsupported input type for function 'cumulative_<operator>': ",
+            batch[0].ToString());
+    }
+
+    RETURN_NOT_OK(builder.FinishInternal(&out_arr));
+    out->value = std::move(out_arr);
+    return Status::OK();
+  }
+
+  static Status Call(KernelContext* ctx, const ArrayData& input,
+                     NumericBuilder<OutType>& builder, ArgValue* accumulator,
+                     bool skip_nulls, bool* encountered_null) {
+    Status st = Status::OK();
+    ArgValue accumulator_tmp = *accumulator;
+    bool encountered_null_tmp = *encountered_null;
+
+    auto null_func = [&]() {
+      st &= builder.AppendNull();
+      encountered_null_tmp = true;
+    };
+
+    if (skip_nulls || (input.GetNullCount() == 0 && !encountered_null_tmp)) {
+      VisitArrayValuesInline<ArgType>(
+          input,
+          [&](ArgValue v) {
+            accumulator_tmp = Op::template Call<OutValue, ArgValue, ArgValue>(
+                ctx, v, accumulator_tmp, &st);
+            st &= builder.Append(accumulator_tmp);
+          },
+          null_func);
+    } else {
+      VisitArrayValuesInline<ArgType>(
+          input,
+          [&](ArgValue v) {
+            if (encountered_null_tmp) {
+              st &= builder.AppendNull();
+            } else {
+              accumulator_tmp = Op::template Call<OutValue, ArgValue, ArgValue>(
+                  ctx, v, accumulator_tmp, &st);
+              st &= builder.Append(accumulator_tmp);
+            }
+          },
+          null_func);
+    }
+
+    *accumulator = accumulator_tmp;
+    *encountered_null = encountered_null_tmp;
+    return st;
+  }
+};
+
+const FunctionDoc cumulative_sum_doc{
+    "Compute the cumulative sum over an array/chunked array of numbers",
+    ("`values` must be an array/chunked array of numeric type values.\n"
```

Comment:
This description states that inputs (`values`) must be an array/chunked array, but the cumulative sum function supports scalars too. Maybe generalize to: "`values` must be numeric.`"

Similarly for the checked variant below.